apps/epl/src/epl_timeline.erl

```diff
@@ -65,34 +66,35 @@ handle_cast({unsubscribe, Pid}, State = #state{subscribers = Subs}) ->
 handle_cast({add, Pid}, State = #state{timelines = Timelines}) when is_pid(Pid) ->
     T = case lists:any(fun({P, _}) -> P == Pid end, Timelines) of
         true -> Timelines;
-        _    ->
-                {ok, Timeline} = epl_timeline_observer:start_link(Pid),
-                [{Pid, Timeline}|Timelines]
+        _    -> [{Pid, track_timeline(Pid)}|Timelines]
         end,
     {noreply, State#state{timelines = T}};
 handle_cast({remove, Pid}, State = #state{timelines = Timelines}) when is_pid(Pid) ->
-    NextTimelines = lists:filter(fun({P, T}) ->
-                                         if
-                                             P == Pid -> exit(T, kill), false;
-                                             true -> true
-                                         end
-                                 end, Timelines),
+    NextTimelines = lists:filter(fun({P, _}) -> P /= Pid end, Timelines),
     {noreply, State#state{timelines = NextTimelines}};
 handle_cast(_Msg, State) ->
     {noreply, State}.
 
-handle_info({data, _, _}, State = #state{subscribers = Subs, timelines = Timelines}) ->
-    T = lists:map(fun({Pid, T}) ->
-                          Timeline = epl_timeline_observer:timeline(T),
-                          #{pid => epl:to_bin(Pid),
-                            timeline => lists:map(fun({M, S}) ->
-                                                          #{message => to_string(M),
-                                                            state => to_string(S)}
-                                                  end, Timeline)}
-                  end, Timelines),
-    JSON = epl_json:encode(#{timelines => T}, <<"timeline-info">>),
+handle_info({data, _, Data}, State = #state{subscribers = Subs, timelines = Times}) ->
+    {Timelines, NewState} =
+        lists:foldl(fun({Pid, {P, PidList, Changes}}, {Acc1, Acc2}) ->
```

Comment:
This anonymous function is huge, so it might be a good idea to move it to private function in the module :)